src/optimizer/cardinality_estimator.cpp

```diff
@@ -0,0 +1,569 @@
+			added_to_eri = true;
+		}
+		// don't add both left and right to the matching_equivalent_sets
+		// since both left and right get added to that index anyway.
+		if (i_set.count(filter_info->right_binding) > 0 && !added_to_eri) {
+			matching_equivalent_sets.push_back(equivalent_relation_index);
+		}
+		equivalent_relation_index += 1;
+	}
+	return matching_equivalent_sets;
+}
+
+void CardinalityEstimator::AddToEquivalenceSets(FilterInfo *filter_info, vector<idx_t> matching_equivalent_sets) {
+	D_ASSERT(matching_equivalent_sets.size() <= 2);
+	if (matching_equivalent_sets.size() > 1) {
+		// an equivalence relation is connecting to sets of equivalence relations
+		// so push all relations from the second set into the first. Later we will delete
+		// the second set.
+		for (ColumnBinding i : relations_to_tdoms.at(matching_equivalent_sets[1]).equivalent_relations) {
+			relations_to_tdoms.at(matching_equivalent_sets[0]).equivalent_relations.insert(i);
+		}
+		relations_to_tdoms.at(matching_equivalent_sets[1]).equivalent_relations.clear();
+		relations_to_tdoms.at(matching_equivalent_sets[0]).filters.push_back(filter_info);
+		// add all values of one set to the other, delete the empty one
+	} else if (matching_equivalent_sets.size() == 1) {
+		idx_t set_i = matching_equivalent_sets.at(0);
+		relations_to_tdoms.at(set_i).equivalent_relations.insert(filter_info->left_binding);
+		relations_to_tdoms.at(set_i).equivalent_relations.insert(filter_info->right_binding);
+		relations_to_tdoms.at(set_i).filters.push_back(filter_info);
+	} else if (matching_equivalent_sets.empty()) {
+		column_binding_set_t tmp;
+		tmp.insert(filter_info->left_binding);
+		tmp.insert(filter_info->right_binding);
+		relations_to_tdoms.emplace_back(RelationsToTDom(tmp));
+		relations_to_tdoms.back().filters.push_back(filter_info);
+	}
+}
+
+void CardinalityEstimator::AddRelationToColumnMapping(ColumnBinding key, ColumnBinding value) {
+	relation_column_to_original_column[key] = value;
+}
+
+void CardinalityEstimator::CopyRelationMap(column_binding_map_t<ColumnBinding> &child_binding_map) {
+	for (auto &binding_map : relation_column_to_original_column) {
+		child_binding_map[binding_map.first] = binding_map.second;
+	}
+}
+
+void CardinalityEstimator::AddColumnToRelationMap(idx_t table_index, idx_t column_index) {
+	relation_attributes[table_index].columns.insert(column_index);
+}
+
+void CardinalityEstimator::InitEquivalentRelations(vector<unique_ptr<FilterInfo>> *filter_infos) {
+	// For each filter, we fill keep track of the index of the equivalent relation set
+	// the left and right relation needs to be added to.
+	vector<idx_t> matching_equivalent_sets;
+	for (auto &filter : *filter_infos) {
+		matching_equivalent_sets.clear();
+
+		if (SingleColumnFilter(filter.get())) {
+			continue;
+		}
+		D_ASSERT(filter->left_set->count >= 1);
+		D_ASSERT(filter->right_set->count >= 1);
+
+		matching_equivalent_sets = DetermineMatchingEquivalentSets(filter.get());
```

Comment:
Nitpick: You can just do `auto matching_equivalent_set = ...` here, don't define it outside the loop, then you don't need to do `matching_equivalent_sets.clear()`.